{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1617e048-2ffc-4370-bc97-a51953a705c6",
   "metadata": {},
   "source": [
    "# Analyze the current Wormcat Annotation List with the Current Wormbase DB\n",
    "\n",
    "1. Find any genes that are on the Wormcat Annotation List but are marked as Dead (Not Live). \n",
    "2. Find any genes that are marked as Live and 'protein coding' but are not on the Wormcat Annotation List. \n",
    "3. Find any genes that have changed their sequence_id.\"\n",
    "\n",
    "Create: `wormcat_genes_for_review.csv`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ddecd934-67ff-4ac8-83b1-4134b7070008",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math\n",
    "import requests\n",
    "import json\n",
    "import csv\n",
    "import time\n",
    "import os\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4657fc4f-d81c-42da-9ca7-b2618babcc6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File exists and will not be pulled.\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Pull Wormcat Category List from source location as needed\n",
    "cd /media/data1/Code/Notebooks/UMass_Med/unknown_genes/input_data\n",
    "\n",
    "WORMCAT_DB=\"whole_genome_v2_nov-11-2021.csv\"\n",
    "WORMCAT_REPO=\" http://www.wormcat.com/static/download\"\n",
    "\n",
    "if [ -f \"$WORMCAT_DB\" ]; then\n",
    "   echo \"File exists and will not be pulled.\"\n",
    "else\n",
    "   echo \"File does not exist locally pulling from repo.\"\n",
    "   wget -q ${WORMCAT_REPO}/${WORMCAT_DB}\n",
    "fi\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "cf8b1db9-30b4-4b2b-80ef-ee20d3a40573",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'WS291'"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the most current Wormbase DB\n",
    "def current_wormbase_version():\n",
    "    api_url = f'http://rest.wormbase.org//rest/database/version'\n",
    "    # Absolutley no error checking is done!!\n",
    "    response = requests.get(api_url)\n",
    "    json_data = json.loads(response.text)\n",
    "    if  response.status_code == 200:\n",
    "        return json_data['data']\n",
    "    else:\n",
    "        return {'error':'somthing is not right'}\n",
    "    \n",
    "current_wormbase_version()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0882f863-8843-41c3-ab72-e2c65a6a2317",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Starting Wormbase Download with version WS291\n",
      "created c_elegans.PRJNA13758.WS291.geneIDs.csv\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "# Pull Wormbase data from source location\n",
    "\n",
    "cd /media/data1/Code/Notebooks/UMass_Med/unknown_genes/input_data\n",
    "\n",
    "#### UPDATE WORMBASE DB ####\n",
    "wormbase_version='WS291'\n",
    "\n",
    "echo \"Starting Wormbase Download with version $wormbase_version\"\n",
    "\n",
    "base_url=\"ftp://ftp.wormbase.org/pub/wormbase/releases/${wormbase_version}/species/c_elegans/PRJNA13758\"\n",
    "gene_ids=\"c_elegans.PRJNA13758.${wormbase_version}.geneIDs.txt.gz\"\n",
    "\n",
    "wget -q ${base_url}/annotation/${gene_ids}\n",
    "gunzip --force ${gene_ids}\n",
    "\n",
    "# Create GeneIDs.csv\n",
    "gene_ids_txt=$(echo \"$gene_ids\" | sed 's/.\\{3\\}$//') # remove .gz\n",
    "gene_ids_csv=$(echo \"$gene_ids\" | sed 's/.\\{7\\}$//') # remove .txt.gz\n",
    "gene_ids_csv=\"${gene_ids_csv}.csv\"                   # add .csv\n",
    "\n",
    "# Drop the first column and Only include Live genes\n",
    "awk -F',' '$5==\"Live\" {print $2\",\"$3\",\"$4\",\"$6}' \"$gene_ids_txt\" > \"$gene_ids_csv\"\n",
    "# Add Header line  \n",
    "sed -i '1iWormbase_Id,Gene_name,Sequence_id,Gene_Type' \"$gene_ids_csv\"\n",
    "echo created $gene_ids_csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "66a1cae4-f4d5-4864-a187-4bf55819e763",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31389\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wc_sequence_id</th>\n",
       "      <th>wormbase_id</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "      <th>category_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F15H10.3</td>\n",
       "      <td>WBGene00000144</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>F35G12.9</td>\n",
       "      <td>WBGene00000145</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C09H10.7</td>\n",
       "      <td>WBGene00007501</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>K06H7.6</td>\n",
       "      <td>WBGene00000143</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>B0511.9</td>\n",
       "      <td>WBGene00015235</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  wc_sequence_id     wormbase_id  category_1       category_2       category_3\n",
       "0       F15H10.3  WBGene00000144  Cell cycle  Cell cycle: APC  Cell cycle: APC\n",
       "1       F35G12.9  WBGene00000145  Cell cycle  Cell cycle: APC  Cell cycle: APC\n",
       "2       C09H10.7  WBGene00007501  Cell cycle  Cell cycle: APC  Cell cycle: APC\n",
       "3        K06H7.6  WBGene00000143  Cell cycle  Cell cycle: APC  Cell cycle: APC\n",
       "4        B0511.9  WBGene00015235  Cell cycle  Cell cycle: APC  Cell cycle: APC"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the Wormbase Annotation List\n",
    "\n",
    "wormcat_df = pd.read_csv('./input_data/whole_genome_v2_nov-11-2021.csv') \n",
    "wormcat_df = wormcat_df.rename(columns={'Sequence ID':'wc_sequence_id','Wormbase ID':'wormbase_id','Category 1':'category_1','Category 2':'category_2','Category 3':'category_3'})\n",
    "wormcat_df = wormcat_df.drop(columns=['Automated Description'])\n",
    "print(len(wormcat_df))\n",
    "wormcat_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "92a7f295-33c8-4699-ad86-7e2e55cd2422",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wormbase_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>wb_sequence_id</th>\n",
       "      <th>status</th>\n",
       "      <th>coding</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00000002</td>\n",
       "      <td>aat-1</td>\n",
       "      <td>F27C8.1</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00000003</td>\n",
       "      <td>aat-2</td>\n",
       "      <td>F07C3.7</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00000004</td>\n",
       "      <td>aat-3</td>\n",
       "      <td>F52H2.2</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00000005</td>\n",
       "      <td>aat-4</td>\n",
       "      <td>T13A10.10</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00000006</td>\n",
       "      <td>aat-5</td>\n",
       "      <td>C55C2.5</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      wormbase_id gene_name wb_sequence_id status               coding\n",
       "0  WBGene00000002     aat-1        F27C8.1   Live  protein_coding_gene\n",
       "1  WBGene00000003     aat-2        F07C3.7   Live  protein_coding_gene\n",
       "2  WBGene00000004     aat-3        F52H2.2   Live  protein_coding_gene\n",
       "3  WBGene00000005     aat-4      T13A10.10   Live  protein_coding_gene\n",
       "4  WBGene00000006     aat-5        C55C2.5   Live  protein_coding_gene"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the wormbase category list\n",
    "\n",
    "gene_ids_df = pd.read_csv('./input_data/c_elegans.PRJNA13758.WS291.geneIDs.txt') \n",
    "#gene_ids_df = wormcat_df.rename(columns={'Sequence ID':'sequence_id','Wormbase ID':'wormbase_id','Category 1':'category_1','Category 2':'category_2','Category 3':'category_3'})\n",
    "gene_ids_df.columns = ['code','wormbase_id','gene_name','wb_sequence_id','status','coding']\n",
    "gene_ids_df = gene_ids_df.drop(columns=['code'])\n",
    "gene_ids_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "47873386-ec6e-4718-86c1-2f8fd71c9c17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wormbase_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>wb_sequence_id</th>\n",
       "      <th>status</th>\n",
       "      <th>coding</th>\n",
       "      <th>wc_sequence_id</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "      <th>category_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WBGene00000002</td>\n",
       "      <td>aat-1</td>\n",
       "      <td>F27C8.1</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>F27C8.1</td>\n",
       "      <td>Transmembrane transport</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WBGene00000003</td>\n",
       "      <td>aat-2</td>\n",
       "      <td>F07C3.7</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>F07C3.7</td>\n",
       "      <td>Transmembrane transport</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>WBGene00000004</td>\n",
       "      <td>aat-3</td>\n",
       "      <td>F52H2.2</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>F52H2.2</td>\n",
       "      <td>Transmembrane transport</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WBGene00000005</td>\n",
       "      <td>aat-4</td>\n",
       "      <td>T13A10.10</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>T13A10.10</td>\n",
       "      <td>Transmembrane transport</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WBGene00000006</td>\n",
       "      <td>aat-5</td>\n",
       "      <td>C55C2.5</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>C55C2.5</td>\n",
       "      <td>Transmembrane transport</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "      <td>Transmembrane transport: amino acid</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      wormbase_id gene_name wb_sequence_id status               coding  \\\n",
       "0  WBGene00000002     aat-1        F27C8.1   Live  protein_coding_gene   \n",
       "1  WBGene00000003     aat-2        F07C3.7   Live  protein_coding_gene   \n",
       "2  WBGene00000004     aat-3        F52H2.2   Live  protein_coding_gene   \n",
       "3  WBGene00000005     aat-4      T13A10.10   Live  protein_coding_gene   \n",
       "4  WBGene00000006     aat-5        C55C2.5   Live  protein_coding_gene   \n",
       "\n",
       "  wc_sequence_id               category_1  \\\n",
       "0        F27C8.1  Transmembrane transport   \n",
       "1        F07C3.7  Transmembrane transport   \n",
       "2        F52H2.2  Transmembrane transport   \n",
       "3      T13A10.10  Transmembrane transport   \n",
       "4        C55C2.5  Transmembrane transport   \n",
       "\n",
       "                            category_2                           category_3  \n",
       "0  Transmembrane transport: amino acid  Transmembrane transport: amino acid  \n",
       "1  Transmembrane transport: amino acid  Transmembrane transport: amino acid  \n",
       "2  Transmembrane transport: amino acid  Transmembrane transport: amino acid  \n",
       "3  Transmembrane transport: amino acid  Transmembrane transport: amino acid  \n",
       "4  Transmembrane transport: amino acid  Transmembrane transport: amino acid  "
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Join the two dataframes\n",
    "\n",
    "merged_df = pd.merge(gene_ids_df, wormcat_df, on='wormbase_id', how='left')\n",
    "merged_df.head()\n",
    "#merged_df.to_csv('merged.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "993b8a6f-045f-48c6-8a9c-cc76ce7666aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wormbase_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>wb_sequence_id</th>\n",
       "      <th>status</th>\n",
       "      <th>coding</th>\n",
       "      <th>wc_sequence_id</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "      <th>category_3</th>\n",
       "      <th>change_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7426</th>\n",
       "      <td>WBGene00007501</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C09H10.7</td>\n",
       "      <td>Dead</td>\n",
       "      <td>gene</td>\n",
       "      <td>C09H10.7</td>\n",
       "      <td>Cell cycle</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "      <td>Cell cycle: APC</td>\n",
       "      <td>Dead Genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7855</th>\n",
       "      <td>WBGene00007930</td>\n",
       "      <td>NaN</td>\n",
       "      <td>C34D1.2</td>\n",
       "      <td>Dead</td>\n",
       "      <td>gene</td>\n",
       "      <td>C34D1.2</td>\n",
       "      <td>Transcription factor</td>\n",
       "      <td>Transcription factor: ZF</td>\n",
       "      <td>Transcription factor: ZF</td>\n",
       "      <td>Dead Genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10089</th>\n",
       "      <td>WBGene00010164</td>\n",
       "      <td>NaN</td>\n",
       "      <td>F56H6.3</td>\n",
       "      <td>Dead</td>\n",
       "      <td>gene</td>\n",
       "      <td>F56H6.3</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Dead Genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14417</th>\n",
       "      <td>WBGene00014492</td>\n",
       "      <td>NaN</td>\n",
       "      <td>R166.6</td>\n",
       "      <td>Dead</td>\n",
       "      <td>gene</td>\n",
       "      <td>R166.6</td>\n",
       "      <td>Non-coding RNA</td>\n",
       "      <td>Non-coding RNA: ncRNA</td>\n",
       "      <td>Non-coding RNA: ncRNA</td>\n",
       "      <td>Dead Genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16267</th>\n",
       "      <td>WBGene00016342</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Dead</td>\n",
       "      <td>gene</td>\n",
       "      <td>C33C12.10</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned</td>\n",
       "      <td>Unassigned: regulated by multiple stresses</td>\n",
       "      <td>Dead Genes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          wormbase_id gene_name wb_sequence_id status coding wc_sequence_id  \\\n",
       "7426   WBGene00007501       NaN       C09H10.7   Dead   gene       C09H10.7   \n",
       "7855   WBGene00007930       NaN        C34D1.2   Dead   gene        C34D1.2   \n",
       "10089  WBGene00010164       NaN        F56H6.3   Dead   gene        F56H6.3   \n",
       "14417  WBGene00014492       NaN         R166.6   Dead   gene         R166.6   \n",
       "16267  WBGene00016342       NaN            NaN   Dead   gene      C33C12.10   \n",
       "\n",
       "                 category_1                category_2  \\\n",
       "7426             Cell cycle           Cell cycle: APC   \n",
       "7855   Transcription factor  Transcription factor: ZF   \n",
       "10089            Unassigned                Unassigned   \n",
       "14417        Non-coding RNA     Non-coding RNA: ncRNA   \n",
       "16267            Unassigned                Unassigned   \n",
       "\n",
       "                                       category_3 change_code  \n",
       "7426                              Cell cycle: APC  Dead Genes  \n",
       "7855                     Transcription factor: ZF  Dead Genes  \n",
       "10089                                  Unassigned  Dead Genes  \n",
       "14417                       Non-coding RNA: ncRNA  Dead Genes  \n",
       "16267  Unassigned: regulated by multiple stresses  Dead Genes  "
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find any Genes that are on the Wormcat Annotation List but are marked as Dead (Not Live)\n",
    "dead_genes_rows = merged_df[(merged_df['status'] != 'Live') & merged_df['category_1'].notna()]\n",
    "dead_genes_rows = dead_genes_rows.copy()\n",
    "dead_genes_rows['change_code']='Dead Genes'\n",
    "print(len(dead_genes_rows))\n",
    "dead_genes_rows.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a42d1ad0-769c-403b-b8fe-825a16f4aec6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "154\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wormbase_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>wb_sequence_id</th>\n",
       "      <th>status</th>\n",
       "      <th>coding</th>\n",
       "      <th>wc_sequence_id</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "      <th>category_3</th>\n",
       "      <th>change_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td>WBGene00000528</td>\n",
       "      <td>clh-1</td>\n",
       "      <td>T27D12.2</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Annotated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>792</th>\n",
       "      <td>WBGene00000859</td>\n",
       "      <td>cwp-1</td>\n",
       "      <td>C37H5.10</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Annotated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>WBGene00002052</td>\n",
       "      <td>ifa-4</td>\n",
       "      <td>K05B2.3</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Annotated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2910</th>\n",
       "      <td>WBGene00002977</td>\n",
       "      <td>lev-10</td>\n",
       "      <td>Y105E8A.7</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Annotated</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3535</th>\n",
       "      <td>WBGene00003602</td>\n",
       "      <td>nhr-3</td>\n",
       "      <td>H01A20.1</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Not Annotated</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         wormbase_id gene_name wb_sequence_id status               coding  \\\n",
       "466   WBGene00000528     clh-1       T27D12.2   Live  protein_coding_gene   \n",
       "792   WBGene00000859     cwp-1       C37H5.10   Live  protein_coding_gene   \n",
       "1985  WBGene00002052     ifa-4        K05B2.3   Live  protein_coding_gene   \n",
       "2910  WBGene00002977    lev-10      Y105E8A.7   Live  protein_coding_gene   \n",
       "3535  WBGene00003602     nhr-3       H01A20.1   Live  protein_coding_gene   \n",
       "\n",
       "     wc_sequence_id category_1 category_2 category_3    change_code  \n",
       "466             NaN        NaN        NaN        NaN  Not Annotated  \n",
       "792             NaN        NaN        NaN        NaN  Not Annotated  \n",
       "1985            NaN        NaN        NaN        NaN  Not Annotated  \n",
       "2910            NaN        NaN        NaN        NaN  Not Annotated  \n",
       "3535            NaN        NaN        NaN        NaN  Not Annotated  "
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find any Genes that are marked as Live and \"protien coding\" but are not on the Wormcat Annotation List\n",
    "not_annotated_rows = merged_df[(merged_df['coding'] == 'protein_coding_gene') & (merged_df['status'] == 'Live') & (merged_df['category_1'].isna())]\n",
    "not_annotated_rows = not_annotated_rows.copy()\n",
    "not_annotated_rows['change_code']='Not Annotated'\n",
    "print(len(not_annotated_rows))\n",
    "not_annotated_rows.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "d32c7eae-ecf6-478e-b27f-48e6346f9b11",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>wormbase_id</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>wb_sequence_id</th>\n",
       "      <th>status</th>\n",
       "      <th>coding</th>\n",
       "      <th>wc_sequence_id</th>\n",
       "      <th>category_1</th>\n",
       "      <th>category_2</th>\n",
       "      <th>category_3</th>\n",
       "      <th>change_code</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10641</th>\n",
       "      <td>WBGene00010716</td>\n",
       "      <td>lge-1</td>\n",
       "      <td>K09C8.4</td>\n",
       "      <td>Live</td>\n",
       "      <td>protein_coding_gene</td>\n",
       "      <td>K08E3.5</td>\n",
       "      <td>Metabolism</td>\n",
       "      <td>Metabolism: carbohydrate</td>\n",
       "      <td>Metabolism: carbohydrate</td>\n",
       "      <td>Sequence Id Changed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          wormbase_id gene_name wb_sequence_id status               coding  \\\n",
       "10641  WBGene00010716     lge-1        K09C8.4   Live  protein_coding_gene   \n",
       "\n",
       "      wc_sequence_id  category_1                category_2  \\\n",
       "10641        K08E3.5  Metabolism  Metabolism: carbohydrate   \n",
       "\n",
       "                     category_3          change_code  \n",
       "10641  Metabolism: carbohydrate  Sequence Id Changed  "
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find any Genes that have changed there sequence_id\n",
    "changed_seq_id_rows = merged_df[(merged_df['wb_sequence_id'].str.lower() != merged_df['wc_sequence_id'].str.lower()) & (merged_df['status'] == 'Live') &merged_df['category_1'].notna()]\n",
    "changed_seq_id_rows = changed_seq_id_rows.copy()\n",
    "changed_seq_id_rows['change_code']='Sequence Id Changed'\n",
    "print(len(changed_seq_id_rows))\n",
    "changed_seq_id_rows.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "31cbc07b-6f68-4a28-b0dd-2a6def816a58",
   "metadata": {},
   "outputs": [],
   "source": [
    "concatenated_df = pd.concat([dead_genes_rows, changed_seq_id_rows, not_annotated_rows], ignore_index=True)\n",
    "column_order = ['wormbase_id', 'change_code', 'status', 'coding', 'gene_name', 'wb_sequence_id', 'wc_sequence_id', 'category_1', 'category_2', 'category_3']\n",
    "concatenated_df = concatenated_df.reindex(columns=column_order)\n",
    "concatenated_df.to_csv('wormcat_genes_for_review.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cbe36d81-e432-4244-9f24-b6dd30b85deb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:dan-dev-sc]",
   "language": "python",
   "name": "conda-env-dan-dev-sc-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
