{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3c84683c-e527-4c8b-91ff-832a3a5219ad",
   "metadata": {},
   "source": [
    "# How do organisms respond to oxidative stress?\n",
    "\n",
    "### Analysis 1\n",
    "* PQ is applied to Wild-type worms\n",
    "* Fastq: Con_N2 vs PQ_N2\n",
    "\n",
    "### Analysis 2\n",
    "* PQ is applied to e113 Worms\n",
    "* Fastq: Con_e113 vs PQ_e113\n",
    "\n",
    "### Analysis 4\n",
    "* Comparing Wild Type and e113 untreated\n",
    "* Fastq: Con_N2 vs Con_e113\n",
    "\n",
    "### Analysis 5\n",
    "* Comparing PQ treated to Wild Type and e113\n",
    "* Fastq: PQ_N2 vs PQ_e113\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "59bbb54a-8783-4d55-a9a3-fa1ba6212656",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from openpyxl import Workbook\n",
    "from openpyxl.styles import PatternFill\n",
    "from openpyxl.utils import get_column_letter\n",
    "\n",
    "\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib_venn import venn2, venn3\n",
    "\n",
    "import seaborn as sns\n",
    "import umap\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e6fc5a16-247e-43d4-9a17-405d4be6dd1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_dir = \"./input_data\"\n",
    "output_dir = \"./output_data\"\n",
    "\n",
    "wormcat_df = pd.read_csv(f\"{input_dir}/whole_genome_v2_nov-11-2021.csv\") \n",
    "\n",
    "\n",
    "analysis_df  = {}\n",
    "for direction in ['up', 'down']:\n",
    "    for analysis in [1, 2, 4, 5]:\n",
    "        df = pd.read_csv(f\"{input_dir}/analysis{analysis}_{direction}-with-correction.csv\") \n",
    "        analysis_df[f\"{direction}_A{analysis}\"] = df\n",
    "        \n",
    "#analysis_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ba9447e7-e5ca-4f59-a03f-b667654aa919",
   "metadata": {},
   "outputs": [],
   "source": [
    "def highlight_rows_by_column(worksheet, column_number, color):\n",
    "    color_fill = PatternFill(start_color=color, end_color=color, fill_type='solid')\n",
    "    df = pd.DataFrame(worksheet.values)\n",
    "    for row_idx, row in enumerate(df.itertuples(), start=1):\n",
    "        if isinstance(row[column_number], bool) and row[column_number] is True:  # Check if the column value is True\n",
    "            for col_idx in range(1, len(row)):\n",
    "                worksheet.cell(row=row_idx, column=col_idx).fill = color_fill\n",
    "\n",
    "def autofit_columns(worksheet):\n",
    "    for column in worksheet.columns:\n",
    "        max_length = 0\n",
    "        column = [cell for cell in column]\n",
    "        for cell in column:\n",
    "            try:\n",
    "                if len(str(cell.value)) > max_length:\n",
    "                    max_length = len(cell.value)\n",
    "            except:\n",
    "                pass\n",
    "        adjusted_width = (max_length + 2) * 1.2  # Adding some extra padding\n",
    "        worksheet.column_dimensions[get_column_letter(column[0].column)].width = adjusted_width\n",
    "\n",
    "def venn_plot(left, right, overlap_set1_set2, label, filename):\n",
    "    label_parts = label.split('_')\n",
    "    analysis_name_map={'A1':'Con_N2 - PQ_N2','A2':'Con_e113 - PQ_e113','A4':'Con_N2 - Con_e113','A5':'PQ_N2 - PQ_e113'}\n",
    "    labels = [analysis_name_map[label_parts[1]],analysis_name_map[label_parts[2]]]\n",
    "    plt.figure(figsize=(7,7))\n",
    "    ax = plt.gca()\n",
    "    v= venn2(subsets=[len(left),len(right),len(overlap_set1_set2)], set_labels=labels, ax=ax,set_colors=('darkviolet','deepskyblue'),alpha=0.5)\n",
    "\n",
    "    # Save the plot\n",
    "    plt.title(f\"{label_parts[0].upper()} Regulated Genes across\\n    {analysis_name_map[label_parts[1]]} and {analysis_name_map[label_parts[2]]}\", fontdict={'fontsize': 18,'fontweight': 'bold'})\n",
    "    plt.savefig(filename)\n",
    "    plt.close()\n",
    "    return\n",
    "\n",
    "def get_value(wormcat_df, key, column_name):\n",
    "    value_at  = wormcat_df.loc[wormcat_df['Wormbase ID'] == key, column_name]\n",
    "    value = \"\"\n",
    "    if not value_at.empty:\n",
    "        value = value_at.iloc[0]\n",
    "    return value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "abf06ee3-8395-47c0-9d47-f7d691b37fd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_ups = [('A1','A2'),('A1','A4'),('A1','A5'),('A2','A4'),('A2','A5'),('A4','A5')]\n",
    "match_up_dfs = {}\n",
    "for direction in ['up', 'down']:  \n",
    "    for match_up in match_ups:\n",
    "        analysis1_df =analysis_df[f\"{direction}_{match_up[0]}\"]\n",
    "        analysis2_df =analysis_df[f\"{direction}_{match_up[1]}\"]\n",
    "        set1 = set(analysis1_df['ID'])\n",
    "        set2 = set(analysis2_df['ID'])\n",
    "        union_set1_set2 = set1.union(set2)\n",
    "        intersect = set1.intersection(set2)\n",
    "        left = set1 - intersect\n",
    "        right = set2 - intersect\n",
    "        data_dict = {}\n",
    "        for key in union_set1_set2:\n",
    "            value = {'left_only'  : True if key in left else False,\n",
    "                     'right_only' : True if key in right else False,\n",
    "                     'intersect'  : True if key in intersect else False,\n",
    "                     'sequence_id': get_value(wormcat_df, key, 'Sequence ID'),\n",
    "                     'category'   : get_value(wormcat_df, key, 'Category 1')\n",
    "                    }\n",
    "            data_dict[key]=value\n",
    "\n",
    "        df = pd.DataFrame.from_dict(data_dict, orient='index')\n",
    "        \n",
    "        label = f\"{direction}_{match_up[0]}_{match_up[1]}\"\n",
    "        match_up_dfs[label]=df\n",
    "        filename = f\"{output_dir}/{label}.png\"\n",
    "        venn_plot(left, right, intersect, label, filename)\n",
    "        \n",
    "#match_up_dfs        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b5a74e7c-f72e-491f-8722-bdb6ce7ee457",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>left_only</th>\n",
       "      <th>right_only</th>\n",
       "      <th>intersect</th>\n",
       "      <th>sequence_id</th>\n",
       "      <th>category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>WBGene00008850</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>F15B9.6</td>\n",
       "      <td>Transmembrane protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00016119</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C25H3.10</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00007531</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C11H1.7</td>\n",
       "      <td>Extracellular material</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00004926</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C08G5.4</td>\n",
       "      <td>Neuronal function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00001462</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>M79.4</td>\n",
       "      <td>Neuronal function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00007345</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C05E7.3</td>\n",
       "      <td>Unassigned</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00022416</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>Y102A11A.6</td>\n",
       "      <td>Trafficking</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00044457</th>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>C18H7.11</td>\n",
       "      <td>Transmembrane protein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00008577</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>F08G2.5</td>\n",
       "      <td>Stress response</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WBGene00000451</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>C33D12.7</td>\n",
       "      <td>Transcription factor</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2101 rows Ã— 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                left_only  right_only  intersect sequence_id  \\\n",
       "WBGene00008850      False       False       True     F15B9.6   \n",
       "WBGene00016119      False        True      False    C25H3.10   \n",
       "WBGene00007531       True       False      False     C11H1.7   \n",
       "WBGene00004926       True       False      False     C08G5.4   \n",
       "WBGene00001462       True       False      False       M79.4   \n",
       "...                   ...         ...        ...         ...   \n",
       "WBGene00007345       True       False      False     C05E7.3   \n",
       "WBGene00022416       True       False      False  Y102A11A.6   \n",
       "WBGene00044457      False        True      False    C18H7.11   \n",
       "WBGene00008577      False       False       True     F08G2.5   \n",
       "WBGene00000451       True       False      False    C33D12.7   \n",
       "\n",
       "                              category  \n",
       "WBGene00008850   Transmembrane protein  \n",
       "WBGene00016119              Unassigned  \n",
       "WBGene00007531  Extracellular material  \n",
       "WBGene00004926       Neuronal function  \n",
       "WBGene00001462       Neuronal function  \n",
       "...                                ...  \n",
       "WBGene00007345              Unassigned  \n",
       "WBGene00022416             Trafficking  \n",
       "WBGene00044457   Transmembrane protein  \n",
       "WBGene00008577         Stress response  \n",
       "WBGene00000451    Transcription factor  \n",
       "\n",
       "[2101 rows x 5 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "match_up_dfs['up_A1_A2']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "739639ea-682d-4544-88d3-9070c762fbea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Put it all together and generate the output\n",
    "output_dir=\"./output_data\"\n",
    "excel_file_path = f\"{output_dir}/compare_across_experiments.xlsx\"\n",
    "if os.path.exists(excel_file_path):\n",
    "    os.remove(excel_file_path)\n",
    "\n",
    "# Create a new workbook\n",
    "workbook = Workbook()\n",
    "\n",
    "# Remove the default \"Sheet\" created by openpyxl\n",
    "default_sheet = workbook['Sheet']\n",
    "workbook.remove(default_sheet)\n",
    "\n",
    "with pd.ExcelWriter(excel_file_path, engine='openpyxl') as writer:\n",
    "    writer.book = workbook\n",
    "    for sheet_name in match_up_dfs.keys():\n",
    "        match_up_dfs[sheet_name] = match_up_dfs[sheet_name].sort_values(by=['left_only', 'right_only', 'intersect', 'category'], ascending=[False, False, False, True])\n",
    "        match_up_dfs[sheet_name].to_excel(writer, sheet_name=sheet_name, index=True, index_label='ID')\n",
    "        \n",
    "        \n",
    "    # Autofit and highlight columns for each sheet\n",
    "    for sheet in writer.sheets.values():\n",
    "        autofit_columns(sheet)\n",
    "        for index, color in  enumerate(['ca7fe8','7fdeff','b3c2ff']):\n",
    "            highlight_rows_by_column(sheet,index+2, color)\n",
    "            \n",
    "         \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "648ddd66-8190-44ca-8e24-3d9c6ba26c8d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:dan-dev-sc]",
   "language": "python",
   "name": "conda-env-dan-dev-sc-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
